core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -2296,13 +2298,41 @@ class ReplicaManager(val config: KafkaConfig,
 
     if (sendZkNotification)
       if (zkClient.isEmpty) {
-        warn("Unable to propagate log dir failure via Zookeeper in KRaft mode")
+        val uuid = logManager.directoryId(dir)
+        if (uuid.isDefined && lifecycleManager.isDefined)
+          lifecycleManager.get.propagateDirectoryFailure(uuid.get)
```

Comment:
When would we ever get into this state. If this can only occur during migration, I thought we agreed that we would fail the broker if a directory failed during migration.

Reply:
I think I'm missing something, why can this only happen during migration? AFAICT this will be called in normal KRaft operation, as long as the line `logManager.handleLogDirFailure(dir)` above does not shutdown the broker due to there being no remaining online directories.
 
We did agree the broker should fail if a directory fails during migration, apologies, I wasn't planning on making that change as part of this PR but I can look into detecting the migration case and addressing that here too.

Reply:
Why would sendZkNotification be enabled in normal KRaft mode?

Reply:
Ah! I think that variable should get a new name. The "Zk" part of it conveys that instead of an RPC to the controller,  the notification goes via ZK, but what the flag actually does is determine whether the controller should know about the failure or not – regardless of what notification mechanism is used. I'll update this PR to rename that variable.